# Data Wrangling with Python: Project (22nd November 2020)

## Overview / Context


As a Data Science Consultant, you have been tasked to provide recommendations on how to reduce the cases of breakdown of buses in the city of New York. 

The dataset that you have been provided comes from the Bus Breakdown and Delay system which collects information from school bus vendors operating out in the field in real time. Bus staff that encounter delays during the route are instructed to radio the dispatcher at the bus vendor’s central office. The bus vendor staff are then instructed to log into the Bus Breakdown and Delay system to record the event and notify OPT. 

OPT customer service agents use this system to inform parents who call with questions regarding bus service. The Bus Breakdown and Delay system is publicly accessible and contains real time updates. All information in the system is entered by school bus vendor staff



## Research Question

Specific research question would then be : <em>**How can we reduce the cases of breakdown of buses in the city of New York?** </em>

Metric for success : The main goal here is **to deduce factors affecting bus breakdowns** we will not be making any predictions and therefore we do not have a specific metric upon which to base our success. Success can be determined by whether or not we get to deduce useful relationships in the data that can help us understand the factors that influence the breakdown of buses in he city of New York.


## Project Deliverable

The expected deliverable for this project will be an Jupyter notebook with data wrangling technique in Python.

## Dataset Explained


The Bus Breakdown and Delays is taken from [here](bit.ly/BusBreakdownDataset) and provided by [data source](https://data.cityofnewyork.us/Transportation/Bus-Breakdown-and-Delays/ez4e-fazm)


## Step 1. Pre-requisites

In [4]:
# Importing pandas library which will help in reading of data from an external 
#source as well and its manipulation.
import pandas as pd

## Step 2. Preparing Dataset


### Step 2.1 Importing Dataset

The dataset on bus breakdowns and delays.

In [5]:
# Load the cab rides dataset from an external csv file and store it in a dataframe called bus_breakdown
bus_breakdown = pd.read_csv('https://bit.ly/BusBreakdownDataset')

Get the structure / glossary the dataframe.

In [10]:
glossary = pd.read_csv('bus_breakdown_delays_glossary.csv', encoding='unicode_escape')
glossary

Column Name  ...    Data Type
0                       School_Year  ...   Plain Text
1                   Busbreakdown_ID  ...       Number
2                          Run_Type  ...   Plain Text
3                            Bus_No  ...   Plain Text
4                      Route_Number  ...   Plain Text
5                            Reason  ...   Plain Text
6                  Schools_Serviced  ...   Plain Text
7                       Occurred_On  ...  Date & Time
8                        Created_On  ...  Date & Time
9                              Boro  ...   Plain Text
10                 Bus_Company_Name  ...   Plain Text
11                 How_Long_Delayed  ...   Plain Text
12    Number_Of_Students_On_The_Bus  ...       Number
13  Has_Contractor_Notified_Schools  ...   Plain Text
14  Has_Contractor_Notified_Parents  ...   Plain Text
15             Have_You_Alerted_OPT  ...   Plain Text
16                      Informed_On  ...  Date & Time
17                  Incident_Number  ...   Plain Text
18                  Last_Updated_On  ...  Date & Time
19        Breakdown_or_Running_Late  ...   Plain Text
20               School_Age_or_PreK  ...   Plain Text

[21 rows x 3 columns]

### Step 2.2 Exploratory Data Analysis

In [15]:
# sampling first 10 records of the breakdown dataset

bus_breakdown.head(10)

School_Year  Busbreakdown_ID  ... Breakdown_or_Running_Late School_Age_or_PreK
0   2015-2016          1227538  ...              Running Late         School-Age
1   2015-2016          1227539  ...              Running Late         School-Age
2   2015-2016          1227540  ...              Running Late              Pre-K
3   2015-2016          1227541  ...              Running Late         School-Age
4   2015-2016          1227542  ...              Running Late         School-Age
5   2015-2016          1227543  ...              Running Late         School-Age
6   2015-2016          1227544  ...              Running Late         School-Age
7   2015-2016          1227545  ...              Running Late         School-Age
8   2015-2016          1227546  ...              Running Late         School-Age
9   2015-2016          1227547  ...              Running Late         School-Age

[10 rows x 21 columns]

In [14]:
# Checking  the last 5 records

bus_breakdown.tail()

School_Year  ...  School_Age_or_PreK
281105   2016-2017  ...               Pre-K
281106   2016-2017  ...               Pre-K
281107   2016-2017  ...          School-Age
281108   2016-2017  ...          School-Age
281109   2016-2017  ...          School-Age

[5 rows x 21 columns]

In [13]:
# Get the total number of records and columns.

bus_breakdown.shape

(281110, 21)

In [20]:
# Check how many breakdows were reported agains Running late

bus_breakdown.Breakdown_or_Running_Late.value_counts()

Running Late    250053
Breakdown        31057
Name: Breakdown_or_Running_Late, dtype: int64

In [26]:
# Why do the breakdowns happen?

bus_breakdown.Reason.value_counts()

Heavy Traffic                  173221
Other                           37579
Mechanical Problem              28162
Won`t Start                     12283
Flat Tire                        8411
Weather Conditions               6932
Late return from Field Trip      5704
Problem Run                      4068
Accident                         2526
Delayed by School                2222
Name: Reason, dtype: int64

In [16]:
# Checking the data types of the dfields.

bus_breakdown.dtypes

School_Year                        object
Busbreakdown_ID                     int64
Run_Type                           object
Bus_No                             object
Route_Number                       object
Reason                             object
Schools_Serviced                   object
Occurred_On                        object
Created_On                         object
Boro                               object
Bus_Company_Name                   object
How_Long_Delayed                   object
Number_Of_Students_On_The_Bus       int64
Has_Contractor_Notified_Schools    object
Has_Contractor_Notified_Parents    object
Have_You_Alerted_OPT               object
Informed_On                        object
Incident_Number                    object
Last_Updated_On                    object
Breakdown_or_Running_Late          object
School_Age_or_PreK                 object
dtype: object

<em><font color = lightblue> There are a total of `281,110` records of bus breakdowns and `21` fields.

The `Last_Updated_on`, `Informed_on`, `Occurred_On` and `Created_On` are saved as objects (text fields) while they represent dates. For accuracy this will need to be converted to dates. 
Further, we notice that there are 8 more times of running late cases than breakdowns, main reason being heavy traffic along the road.
<br><br>The top three reasons for buses running late are Heavy raffic, Mechanical problems and others listed as just others.</font> </em>

## Step 3. Data Preparation.
Our main interest is to interogate the available data so as to deduce any useful relationships, if any, between the dataset features that can help us understand the factors that affect breakdowns and delayed trips for school buses.. <br> Rephrasing this into a research question: **"How can we reduce the cases of breakdown of buses in the city of New York?"**

### 3.1 Data Cleaning

#### 3.1.1 Check to see if we have null entries for records.

In [25]:
#Check to see if we have null records.

bus_breakdown.isna().sum()

School_Year                             0
Busbreakdown_ID                         0
Run_Type                                3
Bus_No                                  9
Route_Number                            7
Reason                                  2
Schools_Serviced                        7
Occurred_On                             0
Created_On                              0
Boro                                13461
Bus_Company_Name                        0
How_Long_Delayed                    35608
Number_Of_Students_On_The_Bus           0
Has_Contractor_Notified_Schools         0
Has_Contractor_Notified_Parents         0
Have_You_Alerted_OPT                    0
Informed_On                             0
Incident_Number                    271627
Last_Updated_On                         0
Breakdown_or_Running_Late               0
School_Age_or_PreK                      0
dtype: int64

In [24]:
# Chaeck for duplicated entries.

bus_breakdown.duplicated().any()

False

<em><font color = lightblue> We notice that there are quite a number of missing entries especially for the `Incident_Number` which are 271,627 entries out of 281,110. These are so many and dropping them would mean loosing significant number of records. It owuld be interesting to know why the cases do not have incident numbers attached tot them (So happens to Boroughs and the delay time)
 <br> <br> Also we note that we do not have any duplicated records in the provided recordset. 

We further notice that the namimg of the dataset columns is uniform and so no much input in this.</font> </em>



### 3.1.2 More focused questions to assist in handing our research question.

**1. Which bus companies that had the highest breakdowns??**

In [27]:
# We make use of the property 'Bus_Company_Name' and use the COUNT()function against this 
# variable for each of its make group. 

bus_breakdown.Bus_Company_Name.value_counts()

G.V.C., LTD.                          19394
LEESEL TRANSPORTATION CORP (B2192)    17200
RELIANT TRANS, INC. (B232             13741
PIONEER TRANSPORTATION CO             12017
BORO TRANSIT, INC.                    11953
                                      ...  
`                                         1
FORTUNA BUS COMPANY                       1
SMART PICK INC                            1
phillip bus service                       1
L&M Bus Corp.                             1
Name: Bus_Company_Name, Length: 117, dtype: int64

<em><font color = lightblue> The leading bus company on the number of cases is `G.V.C., Ltd` with a total of 19,394 cases. followed closely by `Leesel Transportaion Corp (B2192)`with 17,200 cases </font> </em>



---



In [36]:
#  The above can also be achieved by.. 
#
breakdowns = bus_breakdown.groupby(['Bus_Company_Name']).count()
breakdowns

School_Year  ...  School_Age_or_PreK
Bus_Company_Name                         ...                    
1967                                  1  ...                   1
1992                                108  ...                 108
ACME BUS CORP. (B2321)             1602  ...                1602
ADDIES                               26  ...                  26
ALINA SERVICES CORP.                534  ...                 534
...                                 ...  ...                 ...
Y & M TRANSIT CORP (B2321)           28  ...                  28
`                                     1  ...                   1
alina                                 1  ...                   1
gvc                                   5  ...                   5
phillip bus service                   1  ...                   1

[117 rows x 20 columns]

**2. How many students were in the buses when they broke down?**

In [8]:
# Count of the number of students in the bus during breakdown.
#
bus_breakdown.groupby(['Number_Of_Students_On_The_Bus']).count()[['Busbreakdown_ID']].reset_index()

Number_Of_Students_On_The_Bus  Busbreakdown_ID
0                                0           170376
1                                1            15704
2                                2            16737
3                                3            14837
4                                4            12248
..                             ...              ...
219                           6209                1
220                           6219                1
221                           8547                1
222                           9007                1
223                           9052                1

[224 rows x 2 columns]

In [30]:
# Check to see if there are other entried apart from YES / No in the Contactor_Notified_Parents field.

bus_breakdown.Has_Contractor_Notified_Parents.value_counts()

Yes    209632
No      71478
Name: Has_Contractor_Notified_Parents, dtype: int64

In [9]:
# Check to see if there are other entried apart from YES / No in the Contactor_Notified_Parents field.

#This can also be achieved by using `bus_breakdown.Has_Contractor_Notified_Parents.value_counts()`

bus_breakdown['Has_Contractor_Notified_Parents'].nunique()

2



---

